### DISCUSSION TITLE
Does the Delay Effect work?

### DISCUSSION THREAD

**Discussion by Craigiehp** [pschatzmann/arduino-audio-tools#1706] [Q&A] [answered]
Hi,

I have been trying to add an output voice audio delay of approximately 1 second using a i2s microphone and a i2s amp but for the life of me I can't get the Delay Effect to work. I have searched everywhere and feel like I've tried everything. How should I be implementing the Effect?

`#include "AudioTools.h"
 
AudioInfo info(44100, 2, 16);
I2SStream in;
I2SStream out; 

uint16_t duration_ms=1000;
float depth=0.5;
float feedbackAmount=1.0;
uint32_t sampleRate=44100;

Delay dlay(duration_ms, depth, feedbackAmount, sampleRate);

AudioEffectStream effects(dlay);
// audio_tools::AudioEffectStream effects(audio_tools::Delay);
StreamCopy copier(out, effects); 

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); 

  // start I2S in
  Serial.println("starting I2S...");
  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  config_in.port_no = 0;
  config_in.pin_ws = 25;
  config_in.pin_bck = 26;
  config_in.pin_data = 19;
  // config_in.fixed_mclk = sample_rate * 256
  config_in.pin_mck = 3; // must be 0,1 or 3 - only for ESP_IDF_VERSION_MAJOR >= 4

  in.begin(config_in);

  // start I2S out
  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.i2s_format = I2S_STD_FORMAT;
  config_out.is_master = true;
  config_out.port_no = 1;
  config_out.pin_ws = 14;
  config_out.pin_bck = 15;
  config_out.pin_data = 18;

  config_out.pin_mck = 3; // must be 0,1 or 3 - only for ESP_IDF_VERSION_MAJOR >= 4

  out.begin(config_out);

  effects.addEffect(dlay);  
  effects.begin(info);

  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
`

**Comment by pschatzmann** [comment, marked as answer]
There is a [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects)!
